### ISSUE TITLE
Compilation of WarpX fails on Summit when openPMD is enabled and nvcc/11.2 + gcc/10.2 are used

### ISSUE THREAD

**Issue by lucafedeli88** [ECP-WarpX/WarpX#2004]
Hi!

I've just noticed that trying to compile WarpX with `openPMD` enabled and `nvcc/11.2` + `gcc/10.2` results in the following error during the linking phase:

```
[ 99%] Linking CUDA device code CMakeFiles/app.dir/cmake_device_link.o
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIiZNS_15filterParticlesINS_12ParticleTileILi0ELi0ELi4ELi0ENS_20PinnedArenaAllocatorEEENS3_ILi0ELi0ELi4ELi0ENS_16DefaultAllocatorEEEZN16WarpXOpenPMDPlot21WriteOpenPMDParticlesERKNS_6VectorI12ParticleDiagSaISA_EEEEUlRKNS_21ConstParticleTileDataILi0ELi0ELi4ELi0EEEiRKNS_12RandomEngineEE_iiEEDTclsr15particle_detailE6call_ffp1_cvNT0_25ConstParticleTileDataTypeE_ELi0EtlSJ_EEERT_RKSN_OT1_T2_SW_T3_EUliSL_E_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableISN_vEE5valueEvE4typeESQ_OSN_EUlvE_EEvSN_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIiZNS_15filterParticlesINS_12ParticleTileILi0ELi0ELi4ELi0ENS_20PinnedArenaAllocatorEEENS3_ILi0ELi0ELi4ELi0ENS_16DefaultAllocatorEEEZNK19FlushFormatPlotfile14WriteParticlesERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_6VectorI12ParticleDiagSaISI_EEEEUlRKNS_21ConstParticleTileDataILi0ELi0ELi4ELi0EEEiRKNS_12RandomEngineEE_iiEEDTclsr15particle_detailE6call_ffp1_cvNT0_25ConstParticleTileDataTypeE_ELi0EtlSR_EEERT_RKSV_OT1_T2_S14_T3_EUliST_E_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableISV_vEE5valueEvE4typeESY_OSV_EUlvE_EEvSV_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIlZN17ParticleHistogram12ComputeDiagsEiEUliE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OS5_EUlvE_EEvS5_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_6launchIZNS_9ReduceOpsIJNS_11ReduceOpMinENS_11ReduceOpMaxEEE4evalIiNS_10ReduceDataIJddEEEZN15ParticleExtrema12ComputeDiagsEiEUliE_vEEvT_RT0_OT1_EUlvE_EEviimP11CUstream_stOT_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIiZ28filterCopyTransformParticlesILi1ENS_12ParticleTileILi0ELi0ELi4ELi0ENS_16DefaultAllocatorEEES5_iR27PhotonEmissionTransformFuncRK9SmartCopyLi0EET2_RT0_RT1_PSB_SB_OT4_OT3_EUliRKNS_12RandomEngineEE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableISC_vEE5valueEvE4typeET_OSC_EUlvE_EEvSC_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIiZ28filterCopyTransformParticlesILi1ENS_12ParticleTileILi0ELi0ELi4ELi0ENS_16DefaultAllocatorEEES5_iR27PairGenerationTransformFuncRK9SmartCopySA_Li0EET2_RT0_SD_RT1_PSB_SB_SB_OT4_OT5_OT3_EUliRKNS_12RandomEngineEE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableISC_vEE5valueEvE4typeET_OSC_EUlvE_EEvSC_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIiZ28filterCopyTransformParticlesILi1ENS_12ParticleTileILi0ELi0ELi4ELi0ENS_16DefaultAllocatorEEES5_iR20IonizationFilterFuncR23IonizationTransformFuncR9SmartCopyET2_RT0_RT1_SC_OT3_OT5_OT4_EUliRKNS_12RandomEngineEE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableISD_vEE5valueEvE4typeET_OSD_EUlvE_EEvSD_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIlZN23PhotonParticleContainer6PushPXER12WarpXParIterPKNS_9FArrayBoxES7_S7_S7_S7_S7_NS_7IntVectEilliid11ScaleFields6DtTypeEUllE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OSD_EUlvE_EEvSD_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIlZN25PhysicalParticleContainer6PushPXER12WarpXParIterPKNS_9FArrayBoxES7_S7_S7_S7_S7_NS_7IntVectEilliid11ScaleFields6DtTypeEUllE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OSD_EUlvE_EEvSD_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIlZN25PhysicalParticleContainer5PushPEidRKNS_8MultiFabES5_S5_S5_S5_S5_EUllE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OS8_EUlvE_EEvS8_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_14ParallelForRNGIZN25PhysicalParticleContainer9AddPlasmaEiNS_7RealBoxEEUliiiRKNS_12RandomEngineEE_EENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT_vEE5valueEvE4typeERKNS_3BoxEOS9_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIZN25PhysicalParticleContainer9AddPlasmaEiNS_7RealBoxEEUliiiE_EENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoERKNS_3BoxEOS6_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIlZN30RigidInjectedParticleContainer5PushPEidRKNS_8MultiFabES5_S5_S5_S5_S5_EUllE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OS8_EUlvE_EEvS8_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_6launchIZNS_3Gpu5Fuser6LaunchEvEUlvE_EEviimP11CUstream_stOT_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIiZN5WarpX7shiftMFERNS_8MultiFabERKNS_8GeometryEiidbRK16HostDeviceParserILi3EEEUliiiiE0_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoERKNS_3BoxET_OSE_EUlvE_EEvSE_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIiZNK18WarpXLaserProfiles25FieldFunctionLaserProfile14fill_amplitudeEiPKdS5_dPdEUliE_vEENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT0_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoET_OS9_EUlvE_EEvS9_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIZN5WarpX41InitializeExternalFieldsOnGridUsingParserEPNS_8MultiFabES4_S4_RK16HostDeviceParserILi3EES8_S8_iEUliiiE_ZNS2_41InitializeExternalFieldsOnGridUsingParserES4_S4_S4_S8_S8_S8_iEUliiiE0_ZNS2_41InitializeExternalFieldsOnGridUsingParserES4_S4_S4_S8_S8_S8_iEUliiiE1_EENSt9enable_ifIXaaaasr5amrex19MaybeDeviceRunnableIT_vEE5valuesr5amrex19MaybeDeviceRunnableIT0_vEE5valuesr5amrex19MaybeDeviceRunnableIT1_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoERKNS_3BoxESO_SO_OSD_OSE_OSF_EUlvE_EEvSE_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_11ParallelForIZN21MacroscopicProperties34InitializeMacroMultiFabUsingParserEPNS_8MultiFabERK16HostDeviceParserILi3EEiEUliiiE_EENSt9enable_ifIXsr5amrex19MaybeDeviceRunnableIT_vEE5valueEvE4typeERKNS_3Gpu10KernelInfoERKNS_3BoxEOSB_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_6launchIZNS_9ReduceOpsIJNS_11ReduceOpMaxEEE4evalINS_10ReduceDataIJdEEEZN14FieldReduction21ComputeFieldReductionIS3_EEvvEUliiiE_EEvRKNS_3BoxERT_OT0_EUlvE_EEviimP11CUstream_stOT_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_6launchIZNS_9ReduceOpsIJNS_11ReduceOpMinEEE4evalINS_10ReduceDataIJdEEEZN14FieldReduction21ComputeFieldReductionIS3_EEvvEUliiiE_EEvRKNS_3BoxERT_OT0_EUlvE_EEviimP11CUstream_stOT_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
nvlink error   : Entry function '_ZN5amrex13launch_globalILi256EZNS_6launchIZNS_9ReduceOpsIJNS_11ReduceOpSumEEE4evalINS_10ReduceDataIJdEEEZN14FieldReduction21ComputeFieldReductionIS3_EEvvEUliiiE_EEvRKNS_3BoxERT_OT0_EUlvE_EEviimP11CUstream_stOT_EUlvE_EEvT0_' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants
make[2]: *** [CMakeFiles/app.dir/cmake_device_link.o] Error 255

```
I know that the documentation suggests to use `gcc 6.4` + `nvcc 10` on Summit, but I've been using those more up-to-date compilers for a few months without any issue...

**Comment by WeiqunZhang** [comment]
This is a known issue.  You can either use `-Xptxas --disable-optimizer-constants` or disable rdc with `-DAMReX_GPU_RDC=OFF`.  There is a PR for this, #1999.

**Comment by lucafedeli88** [comment]
Thanks!

**Comment by lucafedeli88** [comment]
Should we leave the issue open until https://github.com/ECP-WarpX/WarpX/pull/1999 is merged?

**Comment by ax3l** [comment]
Yes, let's cross link and leave open til merged. Thanks for the report, that's an NVCC 11.2/3 issue.

I wonder if we can also works around at the source, we are using the same filter in plotfiles, but I have not seen the issue there yet.

**Comment by ax3l** [comment]
Just to document this publicly: Note that @maxpkatz recommends not to use
> -Xptxas --disable-optimizer-constants

even if the compiler output suggests this, because it will very likely influence code generation a lot (for the worse).

**Comment by ax3l** [comment]
@lucafedeli88 since you use nvcc 11.2 for a while already, did you see when this was introduced?
I wonder if it's from a change in the parser, since it affects various kernels that support particle filtering.

**Comment by lucafedeli88** [comment]
I think that the last time I compiled WarpX on Summit was a few weeks ago (maybe ~ 2-3 weeks ago)....

**Comment by ax3l** [comment]
![Screenshot from 2021-06-04 21-38-25](https://user-images.githubusercontent.com/1353258/120880132-3672a900-c57d-11eb-97b2-d16601adc465.png)
I wonder if it's from the increased complexity of functions we provided in the Parser, we definitely added a few potentially (register) heavy ones... (e.g. `jn` on device)

We can also save a few registers per kernel by setting `-DWarpX_GPUCLOCK=OFF` (4 registers, 24 bytes cmem according to #1838)

**Comment by ax3l** [comment]
That problem should be fixed with #1999 but I see a regression for NVCC 10.1.243 in #2007

Does Summit's nvcc/11.2 work for you, @lucafedeli88 ?

**Comment by lucafedeli88** [comment]
> That problem should be fixed with #1999 but I see a regression for NVCC 10.1.243 in #2007
> 
> Does Summit's nvcc/11.2 work for you, @lucafedeli88 ?

Yes, I've just compiled WarpX with openPMD  using  `nvcc 11.2` ( `11.2.67` to be precise ) on Summit. I haven't observed any issue.

**Comment by lucafedeli88** [comment]
The issue has been fixed
